File: data-structures/stack/Stack.test.ts
import Stack from './Stack.ts'

function testNewStack() {
    const stack = new Stack<number>()

    console.assert(stack.isEmpty, 'expect Stack<> to be empty')
}

function testStackPush() {
    const stack = new Stack<number>()
    stack.push(1)

    console.assert(!stack.isEmpty, 'expect Stack<> to have some values')
}

function testStackPeek() {
    const stack = new Stack<number>(), expected = 1
    stack.push(expected)
    const value = stack.peek()

    console.assert(!stack.isEmpty, 'expect Stack<> to have some values')
    console.assert(value === expected, `expect to be: ${expected}, got instead: ${value}`)
}

function testStackPop() {
    const stack = new Stack<number>(), expected = 2
    stack.push(1)
    stack.push(expected)

    const value = stack.pop()
    console.assert(value === expected, `expect to be: ${expected}, got instead: ${value}`)

    stack.pop()
    console.assert(stack.isEmpty, 'expect Stack<> to be empty')
}

testNewStack()
testStackPush()
testStackPeek()
testStackPop()

File: data-structures/stack/stack_test.go
package stack

import (
	"testing"
)

func TestStack_IsEmpty(t *testing.T) {
	stack := New()
	if !stack.IsEmpty() {
		t.Error("expect Stack to be empty")
	}
}

func TestStack_Enqueue(t *testing.T) {
	stack := New()

	stack.Push(1)
	if stack.IsEmpty() {
		t.Error("expect Stack to have some values")
	}
}

func TestStack_Dequeue(t *testing.T) {
	stack := New()

	stack.Push(2)
	stack.Push(3)

	if expected, val := 3, stack.Pop(); expected != val {
		t.Errorf("expect pushed value to be [%d], got instead [%d]", expected, val)
	}

	if expected, val := 2, stack.Pop(); expected != val {
		t.Errorf("expect pushed value to be [%d], got instead [%d]", expected, val)
	}

	if !stack.IsEmpty() {
		t.Error("expect Stack to be empty")
	}
}

func TestStack_Peek(t *testing.T) {
	stack := New()

	stack.Push(2)
	stack.Push(3)

	if expected, val := 3, stack.Peek(); expected != val {
		t.Errorf("expect pushed value to be [%d], got instead [%d]", expected, val)
	}

	if expected, val := 3, stack.Peek(); expected != val {
		t.Errorf("expect pushed value to be [%d], got instead [%d]", expected, val)
	}
}
